plugins/woocommerce/includes/class-wc-cache-helper.php

```diff
@@ -26,56 +26,52 @@ class WC_Cache_Helper {
 		add_action( 'wc_ajax_update_order_review', array( __CLASS__, 'update_geolocation_hash' ), 5 );
 		add_action( 'admin_notices', array( __CLASS__, 'notices' ) );
 		add_action( 'delete_version_transients', array( __CLASS__, 'delete_version_transients' ), 10 );
-		add_action( 'wp', array( __CLASS__, 'prevent_caching' ) );
 		add_action( 'clean_term_cache', array( __CLASS__, 'clean_term_cache' ), 10, 2 );
 		add_action( 'edit_terms', array( __CLASS__, 'clean_term_cache' ), 10, 2 );
 	}
 
 	/**
-	 * Set additional nocache headers.
+	 * Prevent caching on certain pages.
 	 *
-	 * @param array $headers Header names and field values.
 	 * @since 3.6.0
+	 *
+	 * @param array<string, string> $headers Header names and field values.
+	 * @return array<string, string> Filtered headers.
 	 */
-	public static function additional_nocache_headers( $headers ) {
-		global $wp_query;
-
-		$agent = isset( $_SERVER['HTTP_USER_AGENT'] ) ? wp_unslash( $_SERVER['HTTP_USER_AGENT'] ) : ''; // phpcs:ignore WordPress.Security.ValidatedSanitizedInput.InputNotSanitized
-
-		$set_cache = false;
-
-		/**
-		 * Allow plugins to enable nocache headers. Enabled for Google weblight.
-		 *
-		 * @param bool $enable_nocache_headers Flag indicating whether to add nocache headers. Default: false.
-		 */
-		if ( apply_filters( 'woocommerce_enable_nocache_headers', false ) ) {
-			$set_cache = true;
+	public static function prevent_caching( $headers ) {
+		if ( ! is_blog_installed() ) {
+			return $headers;
 		}
+		$page_ids = array_filter( array( wc_get_page_id( 'cart' ), wc_get_page_id( 'checkout' ), wc_get_page_id( 'myaccount' ) ) );
 
-		/**
-		 * Enabled for Google weblight.
-		 *
-		 * @see https://support.google.com/webmasters/answer/1061943?hl=en
-		 */
-		if ( false !== strpos( $agent, 'googleweblight' ) ) {
-			// no-transform: Opt-out of Google weblight. https://support.google.com/webmasters/answer/6211428?hl=en.
-			$set_cache = true;
+		if ( ! is_page( $page_ids ) ) {
+			return $headers;
 		}
 
-		if ( false !== strpos( $agent, 'Chrome' ) && isset( $wp_query ) && is_cart() ) {
-			$set_cache = true;
+		self::set_nocache_constants();
+
+		// These directives are all included in `wp_get_nocache_headers()`, with the exclusion of `no-store` for the sake of bfcache.
+		$new_directives = array(
+			// Prevent caching the response in reverse proxies.
+			'private',
+
+			// Ensure freshness of the response but without `no-store` so that bfcache won't be disabled.
+			'no-cache',
+			'must-revalidate',
+			'max-age=0',
+		);
+		$old_directives = array();
+		if ( isset( $headers['Cache-Control'] ) ) {
+			$old_directives = preg_split( '/\s*,\s*/', $headers['Cache-Control'] );
 		}
+		$headers['Cache-Control'] = implode( ', ', array_unique( array_merge( $old_directives, $new_directives ) ) );
 
-		if ( $set_cache ) {
-			$headers['Cache-Control'] = 'no-transform, no-cache, no-store, must-revalidate';
-		}
 		return $headers;
```

Comment:
The PR discussion mentioned adding back `no-store` for authenticated users, at which point are we doing that because I can't seem to see it? unless you're assuming being on cart/checkout/my-account equal being authenticated? which is technically true but doesn't cover all cases of being authenticated.

Reply:
Adding `no-store` or _removing_ the directive? 

Are you referring to [this comment](https://github.com/woocommerce/woocommerce/pull/58445#issuecomment-2935741965)? 

In https://core.trac.wordpress.org/ticket/63636 I'm working on removing `no-store` for authenticated users in core.

Reply:
I was about this comment that includes `no-store` for authed users

https://github.com/woocommerce/woocommerce/pull/58445#issuecomment-2993821380